src/Storage/Device/S3.php

```diff
@@ -169,14 +169,124 @@ public function getPath(string $filename, $prefix = null): string
      *
      * @param string $source
      * @param string $path
+     * @param int chunk
+     * @param int chunks
+     * @param array $metadata
      *
      * @throws \Exception
      *
      * @return int
      */
-    public function upload($source, $path, $chunk = 1, $chunks = 1): int
+    public function upload($source, $path, $chunk = 1, $chunks = 1, &$metadata = []): int
     {
-        return $this->write($path, \file_get_contents($source), \mime_content_type($source));
+        if($chunk == 1 && $chunks == 1) {
+            return $this->write($path, \file_get_contents($source), \mime_content_type($source));
+        }
+        $uploadId = $metadata['uploadId'] ?? null;
+        if(empty($uploadId)) {
+            $uploadId = $this->createMultipartUpload($path, $metadata['content_type']);
+            $metadata['uploadId'] = $uploadId;
+        }
+
+        $etag = $this->uploadPart($source, $path, $chunk, $uploadId);
+        $metadata['parts'] ??= [];
+        $metadata['parts'][] = ['partNumber' => $chunk, 'etag' => $etag];
+        $metadata['chunks'] ??= 0;
+        $metadata['chunks']++;
+        if($metadata['chunks'] == $chunks) {
+            $this->completeMultipartUpload($path, $uploadId, $metadata['parts']);
+        }
+        return $metadata['chunks'];
+    }
+
+    /**
+     * Start Multipart Upload
+     * 
+     * Initiate a multipart upload and return an upload ID.
+     * 
+     * @param string $path
+     * @param string $contentType
+     * 
+     * @return string
+     */
+    protected function createMultipartUpload(string $path, string $contentType): string
+    {
+        $uri = $path !== '' ? '/' . \str_replace(['%2F', '%3F'], ['/', '?'], \rawurlencode($path)) : '/';
+
+        $this->headers['date'] = gmdate('D, d M Y H:i:s T');
+        $this->headers['content-type'] = $contentType;
+        $this->amzHeaders['x-amz-acl'] = $this->acl;
+        $response = $this->call(self::METHOD_POST, $uri, '', ['uploads' => '']);
+        return $response->body['UploadId'];
+    }
+
+    /**
+     * Upload Part
+     * 
+     * @param string $source
+     * @param string $path
+     * @param string $uploadId
+     * @param int $chunk
+     * 
+     * @return string
+     */
+    protected function uploadPart($source, $path, $chunk, $uploadId) : string
+    {
+        $uri = $path !== '' ? '/' . \str_replace(['%2F', '%3F'], ['/', '?'], \rawurlencode($path)) : '/';
+        
+        $data = \file_get_contents($source);
+        $this->headers['date'] = \gmdate('D, d M Y H:i:s T');
+        $this->headers['content-type'] = \mime_content_type($source);
+        $this->headers['content-md5'] = \base64_encode(md5($data, true)); //TODO whould this work well with big file? can we skip it?
```

Comment:
> //TODO whould this work well with big file? can we skip it?

Found this in the comments of `md5()` docs:
> If you want to hash a large amount of data you can use the hash_init/hash_update/hash_final functions.
> This allows you to hash chunks/parts/incremental or whatever you like to call it.

https://www.php.net/manual/en/function.md5.php#94680

Reply:
Thanks, but I think we can skip this for now, as we will be using small chunks for uploading. Any files above 5 MB will be chunked!